easybuild/tools/toolchain/compiler.py

```diff
@@ -278,18 +279,40 @@ def _set_optimal_architecture(self, default_optarch=None):
                                 (--optarch and --optarch=GENERIC still override this value)
         """
         optarch = None
+        use_generic = False
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if build_option('optarch') is not None:
+            for optarch_option in build_option('optarch').split(','):
+                # No compilers specified, and option is not generic, so simply take what is there
+                if ":" not in optarch_option and optarch_option != OPTARCH_GENERIC:
+                    optarch = optarch_option
+                    break
+                # Option is generic, so mark it as to be set later and stop iterating
+                elif optarch_option == OPTARCH_GENERIC:
+                    use_generic = True
+                    break
```

Comment:
I think that being too flexible may be an issue too, since it makes it too easy to have `eb` do something while you're expecting something else. Even if this is clearly documented, we should try and avoid surprises, and do a bit of input validation to protect against silly mistakes (we all make them).

I'm OK with your opposition against using (only) a file, I understand the argument. Also supporting a file can always be added later.

The easiest way to go forward seems to be:

* use different characters than `,` and `:` to specify the mapping of compiler to flags (the `nvcc` flags are an excellent example); wrapping flags in single quotes is not something I'd go for, since you need to be quite careful to make sure bash doesn't eat your single quotes...

I would also input validation, see my previous comment and add to that to *not* allow includes the same compiler twice (like in your `Intel:...,...,Intel:....` example), there's no reason we should allow that, it will only lead to confusion. And it's fairly easy to check.